zeebe/msgpack-value/src/main/java/io/camunda/zeebe/msgpack/value/ObjectValue.java

```diff
@@ -103,7 +103,10 @@ public void read(final MsgPackReader reader) {
       BaseProperty<? extends BaseValue> prop = null;
 
       for (int k = 0; k < declaredProperties.size(); ++k) {
-        final BaseProperty<?> declaredProperty = declaredProperties.get(k);
+        // optimistically start from the same index as the map
+        // we serialize the keys in order, it's our best guess
+        final var index = (i + k) % declaredProperties.size();
+        final BaseProperty<?> declaredProperty = declaredProperties.get(index);
```

Comment:
🙃 It confuses me a bit. We have a list of declared but weren't reading them in order 🤔 


When we have undeclared we always have to check all declared first is this correct? Can we somehow check that we read all declared already? Or that the index is equal to declared, now everything else is undeclared ?

Reply:
I'm not sure about this optimizations as they might lead to incorrect deserialization.
What happens if we add a new property in the "middle". What will an old version decode in that case?
It will decode the last property of the message as "undeclared", while it would have been decoded correctly in this version.

I think we do not need to try to optimize it too much to avoid introducing new bugs, we should stick with incremental improvements.